File: minimal/InJAVA/DemoDevice/org/noname/DemoDevice.java
package org.noname;
import java.util.*;
import org.unirail.BlackBox.Host;
import org.unirail.BlackBox.Host.Pack.Meta.Field;
import org.unirail.BlackBox.Host.Pack.Meta.Field.Bounds;
import static org.unirail.BlackBox.BitUtils.*;
import java.util.concurrent.ConcurrentLinkedQueue;


public class DemoDevice extends Host
{
    public static class HEARTBEAT extends Pack  implements LoopBackDemoChannel.Sendable, LoopBackDemoChannel_ADV.Sendable
    {
        HEARTBEAT() { super(meta, 0); }
        HEARTBEAT(int bytes) { super(meta, bytes); }
        public long custom_mode_GET()//A bitfield for use for autopilot-specific flags
        {  return (get_bytes(data,  0, 4)); }
        public void custom_mode_SET(long  src) //A bitfield for use for autopilot-specific flags
        {  set_bytes((src) & -1L, 4, data,  0); }
        public char base_mode_GET()//System mode bitfield, see MAV_MODE_FLAGS ENUM in mavlink/include/mavlink_types.
        {  return (char)((char) get_bytes(data,  4, 1)); }
        public void base_mode_SET(char  src) //System mode bitfield, see MAV_MODE_FLAGS ENUM in mavlink/include/mavlink_types.
        {  set_bytes((char)(src) & -1L, 1, data,  4); }
        public char mavlink_version_GET()//MAVLink versio
        {  return (char)((char) get_bytes(data,  5, 1)); }
        public void mavlink_version_SET(char  src) //MAVLink versio
        {  set_bytes((char)(src) & -1L, 1, data,  5); }
        public @MAV_TYPE int type_GET()//Type of the MAV (quadrotor, helicopter, etc., up to 15 types, defined in MAV_TYPE ENUM
        {  return  0 + (int)get_bits(data, 48, 5); }
        public void type_SET(@MAV_TYPE int  src) //Type of the MAV (quadrotor, helicopter, etc., up to 15 types, defined in MAV_TYPE ENUM
        {  set_bits(- 0 +   src, 5, data, 48); }
        public @MAV_AUTOPILOT int autopilot_GET()//Autopilot type / class. defined in MAV_AUTOPILOT ENU
        {  return  0 + (int)get_bits(data, 53, 4); }
        public void autopilot_SET(@MAV_AUTOPILOT int  src) //Autopilot type / class. defined in MAV_AUTOPILOT ENU
        {  set_bits(- 0 +   src, 4, data, 53); }
        public @MAV_STATE int system_status_GET()//System status flag, see MAV_STATE ENU
        {  return  0 + (int)get_bits(data, 57, 4); }
        public void system_status_SET(@MAV_STATE int  src) //System status flag, see MAV_STATE ENU
        {  set_bits(- 0 +   src, 4, data, 57); }
        static final Meta meta = new Meta(0, 0, 1, 0, 8, 61);
    }

    public static class LoopBackDemoChannel  extends Channel
    {


        public static  LoopBackDemoChannel instance = new LoopBackDemoChannel();

        public final java.io.InputStream inputStream = new  InputStream();
        //interface-to-mark of sendable through this channel packs_Schs_Rchs
        public interface Sendable {}
        public static HEARTBEAT new_HEARTBEAT() {return new  HEARTBEAT();}

        public void send(Sendable pack) { sendout_packs.add((Pack) pack);}
        protected final Queue<Pack> sendout_packs = new ConcurrentLinkedQueue<Pack>()
        {
            @Override public boolean add(Pack pack)
            {
                synchronized(this)
                {
                    boolean ret = super.add(pack);
                    this.notify();
                    return ret;
                }
            }
        };

        public boolean waitingSendoutPack()
        {
            try
            {
                synchronized(sendout_packs)
                {
                    while(sendout_packs.size() == 0) sendout_packs.wait();
                }
                return true;
            }
            catch(InterruptedException e) {}
            return false;
        }

        public final java.io.OutputStream outputStream = new  OutputStream();
        protected final Queue<Pack> received_packs = new ConcurrentLinkedQueue<Pack>()
        {
            @Override public boolean add(Pack pack)
            {
                synchronized(this)
                {
                    boolean ret = super.add(pack);
                    this.notify();
                    return ret;
                }
            }
        };

        public boolean waitingReceivedPack()
        {
            try
            {
                synchronized(received_packs)
                {
                    while(received_packs.size() == 0) received_packs.wait();
                    return true;
                }
            }
            catch(InterruptedException e) {}
            return false;
        }


        private final Bounds.Inside ph    = new Bounds.Inside();


        @Override protected Pack process(Pack pack, int id)
        {
            for(boolean LOOP = false; ;)
            {
                switch(id)
                {
                    default:
                        assert(false);
                        return null;
                    case 0:
                        if(pack == null) return new HEARTBEAT();
                        final HEARTBEAT pi0 = (HEARTBEAT) pack;
                        ph.setPack(pi0);
                        on_HEARTBEAT.forEach(h -> h.handle(LoopBackDemoChannel.this, ph, pi0));
                        if(LOOP) break;
                        return null;
                    case Channel.PROCESS_CHANNEL_REQEST:
                        if(pack == null) return sendout_packs.poll();
                        received_packs.add(pack);
                        return null;
                    case Channel.PROCESS_HOST_REQEST:
                        if(pack == null) return received_packs.poll();
                        sendout_packs.add(pack);
                        return null;
                    case Channel.PROCESS_RECEIVED:
                        LOOP = true;
                }
                if(received_packs.isEmpty()) return null;
                pack = received_packs.remove();
                id = pack.meta.id;
            }
        }
        public final Collection<OnReceive.Handler<HEARTBEAT, LoopBackDemoChannel>> on_HEARTBEAT = new OnReceive<>();
    }

    public static class LoopBackDemoChannel_ADV  extends Channel
    {


        public static  LoopBackDemoChannel_ADV instance = new LoopBackDemoChannel_ADV();

        public final java.io.InputStream inputStream = new  AdvancedInputStream();
        //interface-to-mark of sendable through this channel packs_Schs_Rchs
        public interface Sendable {}
        public static HEARTBEAT new_HEARTBEAT() {return new  HEARTBEAT();}

        public void send(Sendable pack) { sendout_packs.add((Pack) pack);}
        protected final Queue<Pack> sendout_packs = new ConcurrentLinkedQueue<Pack>()
        {
            @Override public boolean add(Pack pack)
            {
                synchronized(this)
                {
                    boolean ret = super.add(pack);
                    this.notify();
                    return ret;
                }
            }
        };

        public boolean waitingSendoutPack()
        {
            try
            {
                synchronized(sendout_packs)
                {
                    while(sendout_packs.size() == 0) sendout_packs.wait();
                }
                return true;
            }
            catch(InterruptedException e) {}
            return false;
        }


        @Override protected Pack process(Pack pack, int id)
        {
            switch(id)
            {
                default:
                    assert(false);
                    return null;
                case Channel.PROCESS_CHANNEL_REQEST:
                    if(pack == null) return sendout_packs.poll();
                    return null;
                case Channel.PROCESS_HOST_REQEST:
                    sendout_packs.add(pack);
                    return null;
            }
        }

    }

    /*
                  */
    public @interface MAV_TYPE
    {
        int
        MAV_TYPE_GENERIC = 0,
        MAV_TYPE_FIXED_WING = 1,
        MAV_TYPE_QUADROTOR = 2,
        MAV_TYPE_COAXIAL = 3,
        MAV_TYPE_HELICOPTER = 4,
        MAV_TYPE_ANTENNA_TRACKER = 5,
        MAV_TYPE_GCS = 6,
        MAV_TYPE_AIRSHIP = 7,
        MAV_TYPE_FREE_BALLOON = 8,
        MAV_TYPE_ROCKET = 9,
        MAV_TYPE_GROUND_ROVER = 10,
        MAV_TYPE_SURFACE_BOAT = 11,
        MAV_TYPE_SUBMARINE = 12,
        MAV_TYPE_HEXAROTOR = 13,
        MAV_TYPE_OCTOROTOR = 14,
        MAV_TYPE_TRICOPTER = 15,
        MAV_TYPE_FLAPPING_WING = 16;
    }
    /*
    				                                              */
    public @interface MAV_AUTOPILOT
    {
        int
        MAV_AUTOPILOT_GENERIC = 0,
        MAV_AUTOPILOT_PIXHAWK = 1,
        MAV_AUTOPILOT_SLUGS = 2,
        MAV_AUTOPILOT_ARDUPILOTMEGA = 3,
        MAV_AUTOPILOT_OPENPILOT = 4,
        MAV_AUTOPILOT_GENERIC_WAYPOINTS_ONLY = 5,
        MAV_AUTOPILOT_GENERIC_WAYPOINTS_AND_SIMPLE_NAVIGATION_ONLY = 6,
        MAV_AUTOPILOT_GENERIC_MISSION_FULL = 7,
        MAV_AUTOPILOT_INVALID = 8,
        MAV_AUTOPILOT_PPZ = 9,
        MAV_AUTOPILOT_UDB = 10,
        MAV_AUTOPILOT_FP = 11;
    }
    /*
    				                                              */
    public @interface MAV_STATE
    {
        int
        MAV_STATE_UNINIT = 0,
        MAV_STATE_BOOT = 1,
        MAV_STATE_CALIBRATING = 2,
        MAV_STATE_STANDBY = 3,
        MAV_STATE_ACTIVE = 4,
        MAV_STATE_CRITICAL = 5,
        MAV_STATE_EMERGENCY = 6,
        MAV_STATE_POWEROFF = 7;
    }



}

File: minimal/InCS/DemoDevice/DemoDevice.cs

using System;
using System.Collections.Generic;
using System.Threading;
using System.Diagnostics;
using org.unirail.BlackBox;
using Inside = org.unirail.BlackBox.Host.Pack.Meta.Field.Bounds.Inside;
using Field = org.unirail.BlackBox.Host.Pack.Meta.Field;
namespace org.noname
{
    public class DemoDevice : Host
    {
        public class HEARTBEAT : Pack, LoopBackDemoChannel.Sendable, LoopBackDemoChannel_ADV.Sendable
        {
            internal HEARTBEAT() : base(meta0, 0) { }
            internal HEARTBEAT(int bytes) : base(meta0, bytes) { }
            public uint custom_mode //A bitfield for use for autopilot-specific flags
            {
                get {  return (uint)((uint) BitUtils.get_bytes(data,  0, 4));}
                set {  BitUtils.set_bytes((ulong)(value), 4, data,  0);}
            }

            public byte base_mode //System mode bitfield, see MAV_MODE_FLAGS ENUM in mavlink/include/mavlink_types.
            {
                get {  return (byte)((byte) BitUtils.get_bytes(data,  4, 1));}
                set {  BitUtils.set_bytes((ulong)(value), 1, data,  4);}
            }

            public byte mavlink_version //MAVLink versio
            {
                get {  return (byte)((byte) BitUtils.get_bytes(data,  5, 1));}
                set {  BitUtils.set_bytes((ulong)(value), 1, data,  5);}
            }

            public MAV_TYPE type //Type of the MAV (quadrotor, helicopter, etc., up to 15 types, defined in MAV_TYPE ENUM
            {
                get {  return (MAV_TYPE)(0 +  BitUtils.get_bits(data, 48, 5));}
                set {  BitUtils.set_bits((ulong)(- 0 +   value), 5, data, 48);}
            }

            public MAV_AUTOPILOT autopilot //Autopilot type / class. defined in MAV_AUTOPILOT ENU
            {
                get {  return (MAV_AUTOPILOT)(0 +  BitUtils.get_bits(data, 53, 4));}
                set {  BitUtils.set_bits((ulong)(- 0 +   value), 4, data, 53);}
            }

            public MAV_STATE system_status //System status flag, see MAV_STATE ENU
            {
                get {  return (MAV_STATE)(0 +  BitUtils.get_bits(data, 57, 4));}
                set {  BitUtils.set_bits((ulong)(- 0 +   value), 4, data, 57);}
            }
            static readonly Meta meta0 = new Meta(0, 0, 1, 0, 8, 61);
        }

        public class LoopBackDemoChannel : Channel
        {
            public override bool CanRead { get { return true ; } }
            public override bool CanWrite { get { return true; } }
            static LoopBackDemoChannel() {pack_id_bytes = 0;}

            public static  LoopBackDemoChannel instance = new LoopBackDemoChannel();

            //interface-to-mark of sendable through this channel packs_Schs_Rchs
            public interface Sendable {}
            public static HEARTBEAT new_HEARTBEAT() {return new  HEARTBEAT();}

            public void send(Sendable pack) {lock(sendout_packs) {sendout_packs.Enqueue((Pack) pack); Monitor.PulseAll(sendout_packs);}}
            protected readonly Queue<Pack> sendout_packs = new Queue<Pack>();
            protected readonly Queue<Pack> received_packs = new Queue<Pack>();
            private readonly Inside ph    = new Inside();
            protected internal override Pack process(Pack pack, int id)
            {
                for(bool LOOP = false; ;)
                {
                    switch(id)
                    {
                        default:
                            Debug.Assert(false);
                            return null;
                        case 0:
                            if(pack == null) return new HEARTBEAT();
                            if(OnHEARTBEATReceive == null) return null;
                            ph.setPack(pack);
                            OnHEARTBEATReceive(this, ph, (HEARTBEAT) pack);
                            if(LOOP) break;
                            return null;
                        case Channel.PROCESS_CHANNEL_REQEST:
                            if(pack == null) return sendout_packs.Count == 0 ? null : sendout_packs.Dequeue();
                            lock(received_packs) {received_packs.Enqueue(pack); Monitor.PulseAll(received_packs); }
                            return null;
                        case Channel.PROCESS_HOST_REQEST:
                            if(pack == null) return received_packs.Count == 0 ? null : received_packs.Dequeue();
                            lock(sendout_packs) { sendout_packs.Enqueue(pack); Monitor.PulseAll(sendout_packs); }
                            return null;
                        case Channel.PROCESS_RECEIVED:
                            LOOP = true;
                            break;
                    }
                    if(received_packs.Count == 0) return null;
                    pack = received_packs.Dequeue();
                    id = pack.meta.id;
                }
            }
            public event HEARTBEATReceiveHandler OnHEARTBEATReceive;
            public delegate void HEARTBEATReceiveHandler(Channel src, Inside ph, HEARTBEAT pack);
        }
        public class LoopBackDemoChannel_ADV : Channel.Advanced
        {
            public override bool CanRead { get { return false ; } }
            public override bool CanWrite { get { return true; } }
            static LoopBackDemoChannel_ADV() {pack_id_bytes = 0;}

            public static  LoopBackDemoChannel_ADV instance = new LoopBackDemoChannel_ADV();

            //interface-to-mark of sendable through this channel packs_Schs_Rchs
            public interface Sendable {}
            public static HEARTBEAT new_HEARTBEAT() {return new  HEARTBEAT();}

            public void send(Sendable pack) {lock(sendout_packs) {sendout_packs.Enqueue((Pack) pack); Monitor.PulseAll(sendout_packs);}}
            protected readonly Queue<Pack> sendout_packs = new Queue<Pack>();
            protected internal override Pack process(Pack pack, int id)
            {
                switch(id)
                {
                    default:
                        Debug.Assert(false);
                        return null;
                    case Channel.PROCESS_CHANNEL_REQEST:
                        if(pack == null) return sendout_packs.Count == 0 ? null : sendout_packs.Dequeue();
                        return null;
                    case Channel.PROCESS_HOST_REQEST:
                        lock(sendout_packs) { sendout_packs.Enqueue(pack); Monitor.PulseAll(sendout_packs); }
                        return null;
                }
            }
        }
        ; /*
				                                              */

        public enum MAV_TYPE
        {
            MAV_TYPE_GENERIC = 0,
            MAV_TYPE_FIXED_WING = 1,
            MAV_TYPE_QUADROTOR = 2,
            MAV_TYPE_COAXIAL = 3,
            MAV_TYPE_HELICOPTER = 4,
            MAV_TYPE_ANTENNA_TRACKER = 5,
            MAV_TYPE_GCS = 6,
            MAV_TYPE_AIRSHIP = 7,
            MAV_TYPE_FREE_BALLOON = 8,
            MAV_TYPE_ROCKET = 9,
            MAV_TYPE_GROUND_ROVER = 10,
            MAV_TYPE_SURFACE_BOAT = 11,
            MAV_TYPE_SUBMARINE = 12,
            MAV_TYPE_HEXAROTOR = 13,
            MAV_TYPE_OCTOROTOR = 14,
            MAV_TYPE_TRICOPTER = 15,
            MAV_TYPE_FLAPPING_WING = 16
        }
        ; /*
				                                              */

        public enum MAV_AUTOPILOT
        {
            MAV_AUTOPILOT_GENERIC = 0,
            MAV_AUTOPILOT_PIXHAWK = 1,
            MAV_AUTOPILOT_SLUGS = 2,
            MAV_AUTOPILOT_ARDUPILOTMEGA = 3,
            MAV_AUTOPILOT_OPENPILOT = 4,
            MAV_AUTOPILOT_GENERIC_WAYPOINTS_ONLY = 5,
            MAV_AUTOPILOT_GENERIC_WAYPOINTS_AND_SIMPLE_NAVIGATION_ONLY = 6,
            MAV_AUTOPILOT_GENERIC_MISSION_FULL = 7,
            MAV_AUTOPILOT_INVALID = 8,
            MAV_AUTOPILOT_PPZ = 9,
            MAV_AUTOPILOT_UDB = 10,
            MAV_AUTOPILOT_FP = 11
        }
        ; /*
				                                              */

        public enum MAV_STATE
        {
            MAV_STATE_UNINIT = 0,
            MAV_STATE_BOOT = 1,
            MAV_STATE_CALIBRATING = 2,
            MAV_STATE_STANDBY = 3,
            MAV_STATE_ACTIVE = 4,
            MAV_STATE_CRITICAL = 5,
            MAV_STATE_EMERGENCY = 6,
            MAV_STATE_POWEROFF = 7
        }


    }
}